### PR TITLE
Binding, Patterns, and Identifier naming

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

Part of the naming improvements #1819 suggested by @jamiebuilds and a preparation to implement `import`/`export`

* Refines naming of `Binding`s:
  * `JsAnyBinding` is only a `JsIdentifierBinding` or a `JsUnknownBinding`. 
  * Change `id`s to use `JsAnyBinding` instead of `JsUnknownBinding` to allow local error-recovery
  * Rename object and array bindings to `JsArrayBindingPattern` and `JsObjectBindingPattern`
* Refines naming of assignment targets
  * `JsAnyAssignment` now is a union over `StaticMemberAssignment`, `ComputedMemberAssignment`, `UnknownAssignment`, and `IdentifierAssignment`
  * Rename the array and object patterns to `JsArrayAssignmentPattern` and `JsObjectAssignmentPattern`
* Introduce `JsReferenceIdentifier` for identifiers referencing a variable from the current scope. Main reason is to avoid using `IdentifierExpression` in `ShorthandProperties` and `export { name, name2 }` because these names aren't expressions
* Replace `MemberReference` with `JsName`. `JsName` will be used anywhere where an identifier is neither a binding nor a reference (member expression, `import {x as name}`, ...).

### CODE REVIEW

File: crates/rslint_parser/src/syntax/assignment.rs [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+}
+
+fn try_expression_to_assignment(
+	p: &mut Parser,
+	target: CompletedMarker,
+	checkpoint: Checkpoint,
+) -> Result<CompletedMarker, ()> {
+	if !matches!(
+		target.kind(),
+		JS_PARENTHESIZED_EXPRESSION
```
Comment (MichaReiser) [inline crates/rslint_parser/src/syntax/assignment.rs:275]: Hmm, unclear why git doesn't recognise this as the same file as `assignment_target`. I mainly changed this method. I only did renames in the other methods.

File: crates/rslint_parser/src/syntax/stmt.rs [hunk 2]
```diff
@@ -222,50 +221,39 @@ fn parse_expression_statement(p: &mut Parser) -> ParsedSyntax<CompletedMarker> {
...
-					kind: T![ident],
-					range,
-				} => Some(range),
-				_ => None,
-			})
-			.expect(
-				"Tried to get the ident of a name node, but there was no ident. This is erroneous",
-			);
+	// Labelled statement
+	if p.nth_at(1, T![:]) {
```
Comment (MichaReiser) [inline crates/rslint_parser/src/syntax/stmt.rs:225]: No need to first parse an expression and then test if this is a labelled statement. We can simply look if the next token is the colon and if so, treat it as a labelled statement.

**Review by xunilrj — APPROVED** [review]

File: crates/rslint_parser/src/parser.rs [hunk 3]
```diff
@@ -383,7 +388,9 @@ impl<'t> Parser<'t> {
 			errors_pos,
 		} = checkpoint;
 		self.tokens.rewind(token_pos);
-		self.drain_events(self.cur_event_pos() - event_pos);
+		if self.cur_event_pos() > event_pos {
+			self.drain_events(self.cur_event_pos() - event_pos);
+		}
```
Comment (ematipico) [inline crates/rslint_parser/src/parser.rs:393]: What are we fixing here?

File: crates/rslint_parser/src/syntax/expr.rs [hunk 4]
```diff
@@ -1089,14 +1099,18 @@ pub(crate) fn parse_identifier(
...
-				"yield" | "let" if StrictMode.is_supported(p) => Some(
-					p.err_builder(&format!(
-						"Illegal use of `{}` as an identifier in strict mode",
-						name
-					))
-					.primary(p.cur_tok().range, ""),
-				),
-
+				"yield" | "let" | "public" | "protected" | "private" | "package" | "implements"
+				| "interface" | "static"
```
Comment (ematipico) [inline crates/rslint_parser/src/syntax/expr.rs:1103]: `"protected"`, for example, will emit a false negative. Meaning that the usage of that keyword is illegal only in a TypeScript context but JavaScript can use it and it should emit no errors.

We have to revisit the logic and reduce the check only to JS keywords if TS is not enabled.

File: crates/rslint_parser/src/syntax/object.rs [hunk 5]
```diff
@@ -252,8 +244,23 @@ pub(crate) fn parse_object_member_name(p: &mut Parser) -> ParsedSyntax<Completed
 	}
 }
 
+fn is_nth_at_object_member_name(offset: usize, p: &Parser) -> bool {
```
Comment (ematipico) [inline crates/rslint_parser/src/syntax/object.rs:247]:
```rust
fn is_nth_at_object_member_name(p: &Parser, offset: usize) -> bool {
```

nit: I would swap the arguments; most of the times we put the parser in the first place

**Review by ematipico — CHANGES_REQUESTED** [review]

File: crates/rslint_parser/src/parser.rs (hunk 3 above)
Comment (MichaReiser) [reply crates/rslint_parser/src/parser.rs:393]: Let's revert this change. I thought it's needed to use `split_off` in the `rewrite_events` but that's not the case (I must have had another bug in the implementation).

File: crates/rslint_parser/src/syntax/expr.rs (hunk 4 above)
Comment (MichaReiser) [reply crates/rslint_parser/src/syntax/expr.rs:1103]: `protected` is not a valid JS identifier in strict mode

>  It is a Syntax Error if this phrase is contained in strict mode code and the StringValue of IdentifierName is: "implements", "interface", "let", "package", "private", "protected", "public", "static", or "yield". 

Source: https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#sec-identifiers-static-semantics-early-errors

You can also see this from the number of fixed tests

```
xtask/src/coverage/test262/test\language\future-reserved-words\implements-strict.js
xtask/src/coverage/test262/test\language\future-reserved-words\interface-strict.js
xtask/src/coverage/test262/test\language\future-reserved-words\package-strict.js
xtask/src/coverage/test262/test\language\future-reserved-words\private-strict.js
xtask/src/coverage/test262/test\language\future-reserved-words\protected-strict.js
xtask/src/coverage/test262/test\language\future-reserved-words\public-strict.js
xtask/src/coverage/test262/test\language\future-reserved-words\static-strict.js
```

> We have to revisit the logic and reduce the check only to JS keywords if TS is not enabled.

It should be fine even in TS context. We only need to make sure to first eat all accessibility modifiers before calling `parse_identifier` (if we then expect any accessibility modifiers).

Comment (ematipico) [reply crates/rslint_parser/src/syntax/expr.rs:1103]: OK this is new to me 😱

**Review by ematipico — APPROVED** [review]

**Pull request merged**